README.md
# pybmi

Calculate BMI based on height and weight and analyze riskcategory

## Installation

```bash
$ pip install pybmi
```

## Usage

`pybmi` can be used to calculate bmi and identify bmi category and health risk associated to the bmi and provide the count of people in specific category as follows:
from pybmi.pybmi import bmi_analysis
out_stat, bmi_agg = bmi_analysis('/path/input_file.txt')
if out_stat == 0:
    for bmi, bmi_cnt in bmi_agg.items():
        print (bmi + ' : ' + str(bmi_cnt))


## Contributing

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.

## License

`pybmi` was created by Arun Prakash J. It is licensed under the terms of the MIT license.

## Credits

`pybmi` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.1.0 (27/11/2021)

- First release of `pybmi`!

tests/test_bmi.py
from pybmi.bmi import bmi_calculator

out_stat, bmi_val = bmi_calculator(171, 75)
assert out_stat == 0, "Output status doesn't match the expectation"
assert bmi_val == 25.64891761567662, "Incorrect BMI value!"

out_stat, bmi_val = bmi_calculator(2.9241, 75, ht_unit='sqm')
assert out_stat == 0, "Output status doesn't match the expectation"
assert bmi_val == 25.64891761567662, "Incorrect BMI value!"

out_stat, bmi_val = bmi_calculator(171, 150, wt_unit='lb')
assert out_stat == 0, "Output status doesn't match the expectation"
assert bmi_val == 23.26527820525974, "Incorrect BMI value!"

out_stat, bmi_val = bmi_calculator(1.71, 75, ht_unit='m')
assert out_stat == 1, "Output status doesn't match the expectation"
assert bmi_val == 0, "Incorrect BMI value!"

out_stat, bmi_val = bmi_calculator(171, 75, wt_unit='oz')
assert out_stat == 1, "Output status doesn't match the expectation"
assert bmi_val == 0, "Incorrect BMI value!"


tests/test_pybmi.py
from pybmi.pybmi import bmi_analysis


out_stat, bmi_agg = bmi_analysis('c:/test/input_file.txt')
assert bmi_agg == {'Overweight': 1}, "BMI calculation incorrect!"


tests/test_bmicategory.py
from pybmi.bmicategory import BmiCategory

bmi_analysis = BmiCategory()

out_stat, bmi_val = bmi_analysis.get_bmi_cat_risk(4.5)
assert out_stat == 0, "Output status doesn't match the expectation"
assert bmi_val == ('Underweight', 'Malnutrition risk'), "Incorrect bmi category / health risk!"

out_stat, bmi_val = bmi_analysis.get_bmi_cat_risk(18.4)
assert out_stat == 0, "Output status doesn't match the expectation"
assert bmi_val == ('Underweight', 'Malnutrition risk'), "Incorrect bmi category / health risk!"

out_stat, bmi_val = bmi_analysis.get_bmi_cat_risk(18.4)
assert out_stat == 0, "Output status doesn't match the expectation"
assert bmi_val == ('Underweight', 'Malnutrition risk'), "Incorrect bmi category / health risk!"

out_stat, bmi_val = bmi_analysis.get_bmi_cat_risk(18.5)
assert out_stat == 0, "Output status doesn't match the expectation"
assert bmi_val == ('Normal weight', 'Low risk'), "Incorrect bmi category / health risk!"

out_stat, bmi_val = bmi_analysis.get_bmi_cat_risk(24.9)
assert out_stat == 0, "Output status doesn't match the expectation"
assert bmi_val == ('Normal weight', 'Low risk'), "Incorrect bmi category / health risk!"

out_stat, bmi_val = bmi_analysis.get_bmi_cat_risk(25.6)
assert out_stat == 0, "Output status doesn't match the expectation"
assert bmi_val == ('Overweight', 'Enhanced risk'), "Incorrect bmi category / health risk!"

out_stat, bmi_val = bmi_analysis.get_bmi_cat_risk(29.9)
assert out_stat == 0, "Output status doesn't match the expectation"
assert bmi_val == ('Overweight', 'Enhanced risk'), "Incorrect bmi category / health risk!"

out_stat, bmi_val = bmi_analysis.get_bmi_cat_risk(30)
assert out_stat == 0, "Output status doesn't match the expectation"
assert bmi_val == ('Moderately obese', 'Medium risk'), "Incorrect bmi category / health risk!"

out_stat, bmi_val = bmi_analysis.get_bmi_cat_risk(34.9)
assert out_stat == 0, "Output status doesn't match the expectation"
assert bmi_val == ('Moderately obese', 'Medium risk'), "Incorrect bmi category / health risk!"

out_stat, bmi_val = bmi_analysis.get_bmi_cat_risk(35)
assert out_stat == 0, "Output status doesn't match the expectation"
assert bmi_val == ('Severely obese', 'High risk'), "Incorrect bmi category / health risk!"

out_stat, bmi_val = bmi_analysis.get_bmi_cat_risk(39.9)
assert out_stat == 0, "Output status doesn't match the expectation"
assert bmi_val == ('Severely obese', 'High risk'), "Incorrect bmi category / health risk!"

out_stat, bmi_val = bmi_analysis.get_bmi_cat_risk(40)
assert out_stat == 0, "Output status doesn't match the expectation"
assert bmi_val == ('Very Severely obese', 'Very High risk'), "Incorrect bmi category / health risk!"

out_stat, bmi_val = bmi_analysis.get_bmi_cat_risk(70)
assert out_stat == 0, "Output status doesn't match the expectation"
assert bmi_val == ('Very Severely obese', 'Very High risk'), "Incorrect bmi category / health risk!"

out_stat, bmi_val = bmi_analysis.get_bmi_cat_risk(999.9)
assert out_stat == 0, "Output status doesn't match the expectation"
assert bmi_val == ('Very Severely obese', 'Very High risk'), "Incorrect bmi category / health risk!"

out_stat, bmi_val = bmi_analysis.get_bmi_cat_risk(-1)
assert out_stat == 1, "Output status doesn't match the expectation"
assert bmi_val == ('Couldn't fetch health category for - -1', 'Couldn't fetch health category for - -1'), "Incorrect bmi category / health risk!"

out_stat, bmi_val = bmi_analysis.get_bmi_cat_risk(1000.1)
assert out_stat == 1, "Output status doesn't match the expectation"
assert bmi_val == ('Couldn't fetch health category for - 1000.1', 'Couldn't fetch health category for - 1000.1'), "Incorrect bmi category / health risk!"


src/pybmi/bmi.py

def ht_cm_sqm(ht):
    """Converts height from centimeter to squared meter

    Parameters
    ----------
    ht: double
        height in cms

    Returns
    -------
    double
        height in squared meter.

    Examples
    --------
    >>> ht_cm_sqm(175)
    1.75
    """

    return (ht / 100) ** 2


def wt_lb_kg(wt):
    """Converts height from centimeter to squared meter

    Parameters
    ----------
    wt: double
        weight in pound

    Returns
    -------
    double
        weight in kilogram.

    Examples
    --------
    >>> wt_lb_kg(200)
    90.8
    """

    return wt * 0.454


def bmi_calculator(ht, wt, ht_unit='cms', wt_unit='kg'):
    """Calculates BMI

    Before calculating bmi height will be converted to squared meter and weight to kilograms depending

    Parameters
    ----------
    ht: double
    wt: double
    ht_unit: str
        incoming height units - to convert height value into squared meter
        supported units
        ---------------
            cms = centimeters
            sqm = squared meter
    wt_unit: str
        incoming weight units - to convert them into kilogram if its not
        supported units
        ---------------
            kg = kilogram
            lb = pound

    Returns
    -------
    int, double
        int value to represent function successfulness
            0 - success
            1 - failure due to unit value not supported
        bmi value.

    Examples
    --------
    >>> bmi_calculator(3.0625, 75, 'sqm', 'kg')
    0, 24.49

    """

    if ht_unit not in ['cms', 'sqm'] or wt_unit not in ['kg', 'lb']:
        return 1, 0

    ht_sqm = ht

    if ht_unit == 'cms':
        ht_sqm = ht_cm_sqm(ht)

    wt_kg = wt

    if wt_unit == 'lb':
        wt_kg = wt_lb_kg(wt)

    return 0, (wt_kg / ht_sqm)


src/pybmi/pybmi.py
import pathlib
import pandas as pd
import json


import bmi
import bmicategory


def load_data(input_file, read_count):
    """Reads the file chunk based on the read_count and loads the data into a dataframe and return it
    Parameters
    ----------
    input_file: json file name with path
    read_count: int
        data chunk cap

    Returns
    -------
    dataframe
        input file chunk loaded in dataframe
    """
    input_file_data = list()
    counter = 0
    with open(input_file, 'r') as inp_file:
        for line in inp_file:
            line = line.strip(',\n').strip('[').strip(']')
            input_file_data.append(json.loads(line.strip('\n')))
            counter += 1
            if counter == read_count:
                df = pd.DataFrame.from_dict(input_file_data, orient='columns')
                yield df
                counter = 0
                input_file_data.clear()


def add_bmi_health_stat(ht, wt, ht_unit, wt_unit, count_cat):
    """Gets bmi value for each height & weight series element and fetches the bmi category and health risk"""
    bmi_health = dict()
    bmi_stat = dict()
    bmi_cat_risk = bmicategory.BmiCategory()
    for ind, height in ht.items():
        bmi_value = bmi.bmi_calculator(height, wt.at[ind], ht_unit, wt_unit)[1]
        health_check_stat, bmi_cat, health_risk = bmi_cat_risk.get_bmi_cat_risk(round(bmi_value, 1))
        if health_check_stat != 0:
            print('Has issue in getting health category & risk for height: {}, weight: {}, bmi: {}'.format(height,
                                                                                                           wt.at[ind],
                                                                                                           bmi))
        elif bmi_cat in count_cat:
            for cat in count_cat:
                bmi_stat[cat] = bmi_stat.get(cat, 0) + 1

        bmi_health[ind] = {'bmi': bmi_value, 'bmi_category': bmi_cat, 'health_risk': health_risk}

    return bmi_stat, pd.DataFrame.from_dict(bmi_health, orient='index')


def bmi_analysis(input_file, ht_unit='cms', wt_unit='kg', count_cat=['Overweight']):
    """Performs bmi analysis
    Reads from input json (list of dict format) data and calculates bmi and health category and risk
    returns total number of people in provided category(s)

    Parameters
    ----------
    input_file: list of dict / json file
    ht_unit: str
        incoming height units - to convert height value into squared meter
        supported units
        ---------------
            cms = centimeters
            sqm = squared meter
    wt_unit: str
        incoming weight units - to convert them into kilogram if its not
        supported units
        ---------------
            kg = kilogram
            lb = pound
    count_cat: list
        one or more categories for which the function has to return number of people in the specified category

    Returns
    -------
    int, dict
        int value to represent function successfulness
            0 - success
            1 - failure due to unit value not supported
        dictionary containing count of provided bmi categories.
    """

    file = pathlib.Path(input_file)

    if file.exists():
        proceed = True
    else:
        proceed = False
        print('{} file does not exist'.format(input_file))

    if proceed:
        file_data = load_data(input_file, 30)
        data_df = list()
        bmi_stat_out = dict()
        while True:
            try:
                input_df = next(file_data)
                ht = input_df['HeightCm']
                wt = input_df['WeightKg']
                temp_bmi_stat, bmi_cat_risk = add_bmi_health_stat(ht, wt, ht_unit, wt_unit, count_cat)

                for key_val, key_count in temp_bmi_stat.items():
                    bmi_stat_out[key_val] = bmi_stat_out.get(key_val, 0) + key_count

                bmi_out = pd.merge(input_df, bmi_cat_risk, left_index=True, right_index=True)
                data_df.append(bmi_out)
            except StopIteration as e:
                print('Completed iterating over the input file' + str(e))
                return 0, bmi_stat_out
    else:
        return 1, {}



src/pybmi/bmicategory.py
import pandas as pd


class BmiCategory:
    """PyBmiCategory class - acts as an interface to fetch health category & risk for the passed on bmi value"""

    def __init__(self):
        """Initialize bmi category and risk data as dataframe"""

        bmi_category_risk = [{"category": "Underweight", "bmi_min_range": 0.0, "bmi_max_range": 18.4,
                              "health_risk": "Malnutrition risk"},
                             {"category": "Normal weight", "bmi_min_range": 18.5, "bmi_max_range": 24.9,
                              "health_risk": "Low risk"},
                             {"category": "Overweight", "bmi_min_range": 25, "bmi_max_range": 29.9,
                              "health_risk": "Enhanced risk"},
                             {"category": "Moderately Obese", "bmi_min_range": 30, "bmi_max_range": 34.9,
                              "health_risk": "Medium risk"},
                             {"category": "Severely Obese", "bmi_min_range": 35, "bmi_max_range": 39.9,
                              "health_risk": "High risk"},
                             {"category": "Very Severely Obese", "bmi_min_range": 40, "bmi_max_range": 999.9,
                              "health_risk": "Very High risk"}]

        self.bmi_cat_risk_df = pd.DataFrame.from_dict(bmi_category_risk, orient='columns')

    def get_bmi_cat_risk(self, bmi_value):
        """Provides health category and risk based on provided BMI value

        Parameters
        ----------
        self: onject of PyBmiCategory
        bmi_value: double
            Provided bmi value

        Returns
        -------
        int, str, str
            int value to represent function successfulness
                0 - success
                1 - failure due to unit value not supported
            first string value provides health category
            second string value provides health risk

        Examples
        --------
        >>> get_bmi_cat_risk(24.489795918367346)
        0, "Normal weight", "Low risk"

        """

        # global bmi_category_risk
        # bmi_cat_risk_df = pd.DataFrame.from_dict(bmi_category_risk, orient='columns')
        bmi_value_out = self.bmi_cat_risk_df[(self.bmi_cat_risk_df['bmi_min_range'] <= bmi_value)
                                             & (self.bmi_cat_risk_df['bmi_max_range'] >= bmi_value)]
        if bmi_value_out.count()[0] > 0:
            return 0, \
                   bmi_value_out['category'].values[0], \
                   bmi_value_out['health_risk'].values[0]
        else:
            return 1, \
                   "Couldn't fetch health category for - " + str(bmi_value), \
                   "Couldn't fetch health risk for - " + str(bmi_value)
